### PR TITLE
Optional change address

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x] adjust post wallet api and enabling one change addr flag
- [x] extend SeqState
- [x] adjust nextChangeIndex
- [x] update db schema and related code
- [x] handle db migration
- [x] add integration test proving the case
- [x] allow wallet to go from multi-change-address to one-change-address mode and vice versa (update wallet api change)
- [x] show the case in integration testing
- [x] add spec
- [x] add PUT for shared wallets
- [x] do the same like for shelley for shared wallet   

### Issue Number

ADP-3145

### CODE REVIEW

**Review by HeinrichApfelmus — COMMENTED** [review]
Thank you! 😊 I have finally found the time to take a first look.

As for the user-facing design, I'm a bit unsure whether a state flag `oneChangeAddressMode` is the best solution — one alternative design would be to allow the user to supply an explicit change address in the "balance transaction" endpoint. That said, while this alternative has the advantage of avoiding more state, I'm not sure about it either. 🤔 In the end, I agree the state flag is reasonable.

Concerning the implementation, I have three larger requests:

1. Instead of using `Bool`, could you use a more "speaking" enumeration type? I.e.

```
data SeqState … = { …
    , changeAddressMode :: ChangeAddressMode
    }

data ChangeAddressMode
    = SingleChangeAddress
    | IncreasingChangeAddresses
```

This makes the code easier to understand and maintain in the long run.

2. For the implementation of `setOneChangeAddressMode` and `setOneChangeAddressModeShared`, we are now in the excellent position that we can work with the `WalletState` data type directly — it captures the _entire_ state of the wallet in an ordinary Haskell data type, the state is no longer implicit in the database.

   I have taken care of this point and added a commit which rewrites these two functions as pure functions with a little glue code `onWalletState`.

3. For the database, please use the `Migration.New` facilities — the old way is dead; the new way is much more robust. You will need to implement a migration and add it to `runNewStyleMigrations`. Implementing a new `Migration` is straightforward, see for example `migrateDelegations` in  https://github.com/cardano-foundation/cardano-wallet/blob/6986d98fcb029c4142d3c553e5b4541aaa33e331/lib/wallet/src/Cardano/Wallet/DB/Store/Delegations/Migration.hs#L94-L110

But other than that, I think this looks good.

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery.hs [hunk 1]
```diff
@@ -156,6 +161,17 @@ class GetPurpose (key :: Depth -> Type -> Type)  where
 class GetAccount s (key :: Depth -> Type -> Type) | s -> key  where
     getAccount :: s -> key 'AccountK XPub
 
+data ChangeAddressMode
+    = SingleChangeAddress
+    | IncreasingChangeAddresses
+    deriving stock (Generic, Show, Eq)
```
Comment (HeinrichApfelmus) [inline lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery.hs]:
```suggestion
-- | How to generate change addresses.
data ChangeAddressMode
    = SingleChangeAddress
        -- ^ Use a single address for all change outputs.
    | IncreasingChangeAddresses
        -- ^ For every change output, increase a counter and derive an address from that.
    deriving stock (Generic, Show, Eq)
```

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs [hunk 2]
```diff
@@ -334,6 +336,10 @@ data SeqState (n :: NetworkDiscriminant) k = SeqState
         -- ^ Reward account public key associated with this wallet
     , derivationPrefix :: DerivationPrefix
         -- ^ Derivation path prefix from a root key up to the internal account
+    , oneChangeAddressMode :: ChangeAddressMode
+        -- ^ One change address mode. If switched on then next transactions will
+        -- use the same change address. If no then next change index for
+        -- each next transaction is used
```
Comment (HeinrichApfelmus) [inline lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs]:
```suggestion
    , changeAddressMode :: ChangeAddressMode
        -- ^ How to generate change addresses.
```

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs [hunk 3]
```diff
@@ -413,15 +419,17 @@ mkSeqStateFromAccountXPub
...
-mkSeqStateFromAccountXPub accXPub policyXPubM purpose g = SeqState
+mkSeqStateFromAccountXPub accXPub policyXPubM purpose g change = SeqState
     { internalPool = newSeqAddressPool @n accXPub g
     , externalPool = newSeqAddressPool @n accXPub g
     , accountXPub = accXPub
     , policyXPub = policyXPubM
     , rewardAccountKey = rewardXPub
     , pendingChangeIxs = emptyPendingIxs
     , derivationPrefix = DerivationPrefix ( purpose, coinTypeAda, minBound )
+    , oneChangeAddressMode = change
```
Comment (HeinrichApfelmus) [inline lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs]:
```suggestion
    , changeAddressMode = change
```

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs [hunk 4]
```diff
@@ -503,8 +509,14 @@ instance
...
       where
+        ixMin = minBound @(Index 'Soft 'CredFromKeyK)
+        updatePending pendingIxs =
+            pendingIxsFromList $ L.nub $ (pendingIxsToList pendingIxs) <> [ixMin]
         (ix, pending') =
-            nextChangeIndex (getPool $ internalPool st) (pendingChangeIxs st)
+            if oneChangeAddressMode st == SingleChangeAddress  then
+                ( ixMin, updatePending (pendingChangeIxs st) )
+            else
+                nextChangeIndex (getPool $ internalPool st) (pendingChangeIxs st)
```
Comment (HeinrichApfelmus) [inline lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs]:
```suggestion
            case changeAddressMode st of
                SingleChangeAddress ->
                    ( ixMin, updatePending (pendingChangeIxs st) )
                IncreasingChangeAddresses ->
                    nextChangeIndex (getPool $ internalPool st) (pendingChangeIxs st)
```

A `case` statement "speaks" more, making it clearer which case is supposed to do what. 🤓

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Shared.hs [hunk 5]
```diff
@@ -275,6 +277,10 @@ data SharedState (n :: NetworkDiscriminant) k = SharedState
         -- ^ Reward account script hash associated with this wallet
     , poolGap :: !AddressPoolGap
         -- ^ Address pool gap to be used in the address pool of shared state
+    , oneChangeAddressMode :: ChangeAddressMode
+        -- ^ One change address mode. If switched on then next transactions will
+        -- use the same change address. If no then next change index for
+        -- each next transaction is used
```
Comment (HeinrichApfelmus) [inline lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Shared.hs]:
```suggestion
    , changeAddressMode :: ChangeAddressMode
        -- ^ How to generate change addresses.
```

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Shared.hs [hunk 6]
```diff
@@ -511,8 +519,15 @@ instance GenChange (SharedState n k) where
...
-            let (ix, pending') = nextChangeIndex (getPool intPool) pending
+        Active pools@(SharedAddressPools extPool intPool pending) ->
+            let ixMin = minBound @(Index 'Soft 'CredFromScriptK)
+                updatePending pendingIxs =
+                    pendingIxsFromList $ L.nub $ (pendingIxsToList pendingIxs) <> [ixMin]
+                (ix, pending') =
+                    if oneChangeAddressMode st == SingleChangeAddress then
+                        ( ixMin, updatePending (pendingChangeIxs pools) )
+                    else
+                        nextChangeIndex (getPool intPool) pending
```
Comment (HeinrichApfelmus) [inline lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Shared.hs]:
```suggestion
                    case changeAddressMode st of
                        SingleChangeAddress ->
                            ( ixMin, updatePending (pendingChangeIxs pools) )
                        IncreasingChangeAddresses ->
                            nextChangeIndex (getPool intPool) pending
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 7]
```diff
@@ -892,7 +895,8 @@ postShelleyWallet
     -> Handler ApiWallet
 postShelleyWallet ctx generateKey body = do
     let state = mkSeqStateFromRootXPrv
-            (keyFlavorFromState @s) (RootCredentials rootXPrv pwdP) purposeCIP1852 g
+            (keyFlavorFromState @s) (RootCredentials rootXPrv pwdP)
+            purposeCIP1852 g oneAddrMode
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
            purposeCIP1852 g changeAddrMode
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 8]
```diff
@@ -913,6 +917,9 @@ postShelleyWallet ctx generateKey body = do
     wid = WalletId $ digest ShelleyKeyS $ publicKey ShelleyKeyS rootXPrv
     wName = getApiT (body ^. #name)
     genesisParams = ctx ^. #netParams
+    oneAddrMode = case body ^. #oneChangeAddressMode of
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
    changeAddrMode = case body ^. #oneChangeAddressMode of
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 9]
```diff
@@ -1119,6 +1127,9 @@ postSharedWalletFromRootXPrv ctx generateKey body = do
     scriptValidation =
         maybe RecommendedValidation getApiT (body ^. #scriptValidation)
     genesisParams = ctx ^. #netParams
+    oneAddrMode = case body ^. #oneChangeAddressMode of
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
    changeAddrMode = case body ^. #oneChangeAddressMode of
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 10]
```diff
@@ -1080,7 +1087,8 @@ postSharedWalletFromRootXPrv ctx generateKey body = do
     ix' <- liftHandler $ withExceptT ErrConstructSharedWalletInvalidIndex $
         W.guardHardIndex ix
     let state = mkSharedStateFromRootXPrv kF
-            (RootCredentials rootXPrv pwdP) ix' g pTemplate dTemplateM
+            (RootCredentials rootXPrv pwdP) ix' oneAddrMode
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
            (RootCredentials rootXPrv pwdP) ix' changeAddrMode
```

File: lib/wallet/src/Cardano/Wallet.hs [hunk 11]
```diff
@@ -3169,6 +3173,30 @@ writePolicyPublicKey ctx wid pwd = db & \DBLayer{..} -> do
...
+     . s ~ SharedState n SharedKey
+    => WalletLayer IO s
+    -> ChangeAddressMode
+    -> IO ()
+setOneChangeAddressModeShared ctx modeOnOff =
+    onWalletState ctx $ update $ \s ->
+        let (SharedPrologue sharedState) = WS.prologue s
+            sharedState' = sharedState & #oneChangeAddressMode .~ modeOnOff
+        in  [ReplacePrologue $ SharedPrologue sharedState']
+
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet.hs]:
```suggestion
setChangeAddressMode
    :: forall s n
     . s ~ SeqState n ShelleyKey
    => WalletLayer IO s
    -> ChangeAddressMode
    -> IO ()
setChangeAddressMode ctx mode =
    onWalletState ctx $ update $ \s ->
        let (SeqPrologue seqState) = WS.prologue s
            seqState' = seqState & #changeAddressMode .~ mode
        in  [ReplacePrologue $ SeqPrologue seqState']

setChangeAddressModeShared
    :: forall s n
     . s ~ SharedState n SharedKey
    => WalletLayer IO s
    -> ChangeAddressMode
    -> IO ()
setChangeAddressModeShared ctx mode =
    onWalletState ctx $ update $ \s ->
        let (SharedPrologue sharedState) = WS.prologue s
            sharedState' = sharedState & #changeAddressMode .~ mode
        in  [ReplacePrologue $ SharedPrologue sharedState']

```

File: lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs [hunk 12]
```diff
@@ -95,18 +95,20 @@ mkSharedStateFromAccountXPub
     => KeyFlavorS k
     -> k 'AccountK XPub
     -> Index 'Hardened 'AccountK
+    -> ChangeAddressMode
     -> AddressPoolGap
     -> ScriptTemplate
     -> Maybe ScriptTemplate
     -> SharedState n k
-mkSharedStateFromAccountXPub kF accXPub accIx gap pTemplate dTemplateM =
+mkSharedStateFromAccountXPub kF accXPub accIx modeOnOff gap pTemplate dTemplateM =
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs]:
```suggestion
mkSharedStateFromAccountXPub kF accXPub accIx mode gap pTemplate dTemplateM =
```

File: lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs [hunk 13]
```diff
@@ -95,18 +95,20 @@ mkSharedStateFromAccountXPub
...
     activate kF $ SharedState
         { derivationPrefix = DerivationPrefix (purposeCIP1854, coinTypeAda, accIx)
         , accountXPub = accXPub
         , paymentTemplate = pTemplate
         , delegationTemplate = dTemplateM
         , rewardAccountKey = Nothing
         , poolGap = gap
+        , oneChangeAddressMode = modeOnOff
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs]:
```suggestion
        , oneChangeAddressMode = mode
```

File: lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs [hunk 14]
```diff
@@ -116,12 +118,13 @@ mkSharedStateFromRootXPrv
...
     -> Index 'Hardened 'AccountK
+    -> ChangeAddressMode
     -> AddressPoolGap
     -> ScriptTemplate
     -> Maybe ScriptTemplate
     -> SharedState n k
-mkSharedStateFromRootXPrv kF (RootCredentials rootXPrv pwd) accIx =
-    mkSharedStateFromAccountXPub kF accXPub accIx
+mkSharedStateFromRootXPrv kF (RootCredentials rootXPrv pwd) accIx modeOnOff =
+    mkSharedStateFromAccountXPub kF accXPub accIx modeOnOff
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs]:
```suggestion
mkSharedStateFromRootXPrv kF (RootCredentials rootXPrv pwd) accIx mode =
    mkSharedStateFromAccountXPub kF accXPub accIx mode
```

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/New.hs [hunk 15]
```diff
@@ -60,15 +62,20 @@ getVersionNew = fmap oldToNewSchemaVersion . getSchemaVersion
...
 _useSqlBackend = hoistMigration $ withReaderT dbBackend
 
 runNewStyleMigrations :: Tracer IO DBLog -> FilePath -> IO ()
-runNewStyleMigrations tr fp =
+runNewStyleMigrations tr fp = do
+    runMigrations (newMigrationInterface tr) fp
+        $ migrateDelegations . noMigrations2
     runMigrations (newMigrationInterface tr) fp
-        $ migrateDelegations . noMigrations
+        $ migratePrologue . noMigrations3
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/New.hs]:
```suggestion
        $ migratePrologue . migrateDelegations . noMigrations2
```

The intention is that we use `(.)` as composition operator. 🤓 In this way, we get a static check that the migrations connect to each other, and that we have not forgotten any.

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/New.hs [hunk 16]
```diff
@@ -60,15 +62,20 @@ getVersionNew = fmap oldToNewSchemaVersion . getSchemaVersion
...
 _useSqlBackend
     :: Migration (SqlPersistT m) from to
     -> Migration (ReadDBHandle m) from to
 _useSqlBackend = hoistMigration $ withReaderT dbBackend
 
 runNewStyleMigrations :: Tracer IO DBLog -> FilePath -> IO ()
-runNewStyleMigrations tr fp =
+runNewStyleMigrations tr fp = do
+    runMigrations (newMigrationInterface tr) fp
+        $ migrateDelegations . noMigrations2
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/New.hs]:
```suggestion
runNewStyleMigrations tr fp =
```

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Schema.hs [hunk 17]
```diff
@@ -346,6 +347,7 @@ SeqState
     seqStatePolicyXPub        B8.ByteString Maybe  sql=policy_xpub
     seqStateRewardXPub        B8.ByteString        sql=reward_xpub
     seqStateDerivationPrefix  W.DerivationPrefix   sql=derivation_prefix
+    seqStateOneChangeAddrMode W.ChangeAddressMode  sql=change_addr_mode
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Sqlite/Schema.hs]:
```suggestion
    seqStateChangeAddrMode W.ChangeAddressMode  sql=change_addr_mode
```

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Schema.hs [hunk 18]
```diff
@@ -433,6 +435,7 @@ SharedState
     sharedStateDelegationScript         (Script Cosigner) Maybe   sql=delegation_script
     sharedStateRewardAccount            W.RewardAccount Maybe     sql=reward_account
     sharedStateDerivationPrefix         W.DerivationPrefix        sql=derivation_prefix
+    sharedStateOneChangeAddrMode        W.ChangeAddressMode       sql=change_addr_mode
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Sqlite/Schema.hs]:
```suggestion
    sharedStateChangeAddrMode        W.ChangeAddressMode       sql=change_addr_mode
```

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/Old.hs [hunk 19]
```diff
@@ -779,6 +780,15 @@ migrateManually tr key defaultFieldValues =
...
 
+    -- | Adds an 'one_change_addr_mode' column to the 'seq_state'
+    -- table if it is missing.
+    --
+    addOneChangeAddrModeIfMissing :: Sqlite.Connection -> IO ()
+    addOneChangeAddrModeIfMissing conn = do
+        addColumn_ conn True (DBField SeqStateOneChangeAddrMode) value
+      where
+        value = "FALSE"
+
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/Old.hs]:
```suggestion
```

This goes away now.

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/Old.hs [hunk 20]
```diff
@@ -170,6 +170,7 @@ migrateManually tr key defaultFieldValues =
     , moveRndUnusedAddresses
     , cleanupSeqStateTable
     , addPolicyXPubIfMissing
+    , addOneChangeAddrModeIfMissing
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/Old.hs]:
```suggestion
```

File: lib/wallet/src/Cardano/Wallet/DB/Store/WalletState/Migration.hs
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Store/WalletState/Migration.hs:1]: Could you please move this module to `Cardano.Wallet.DB.Store.Checkpoints.Migration`? That puts it closer to the thing that is being migrated (i.e. the address state).

**Review by HeinrichApfelmus — COMMENTED** [review]
Thanks! 😊 I have a couple more requests:

1. With the "speaking" constructors for `ChangeAddressMode`, the name `oneChangeAddresMode` no longer makes sense — `changeAddresMode` is better. I had change it in many places, but not all of them — could you complete this?
2. A couple of requests on the use of the `Migration.New` facilities. In particular, could you move the module where the migration is define to the `Checkpoints.Migration`?

I'll take another look at the documentation after that.

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery.hs]: done

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs]: done

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs]: done

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs (hunk 4 above)
Comment (paweljakubas) [reply lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Sequential.hs]: done

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Shared.hs (hunk 5 above)
Comment (paweljakubas) [reply lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Shared.hs]: done

File: lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Shared.hs (hunk 6 above)
Comment (paweljakubas) [reply lib/address-derivation-discovery/lib/Cardano/Wallet/Address/Discovery/Shared.hs]: done

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 8 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: done

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 9 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: done

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 10 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: done

File: lib/wallet/src/Cardano/Wallet.hs (hunk 11 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet.hs]: done

File: lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs (hunk 13 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs]: done

File: lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs (hunk 14 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs]: done

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/New.hs (hunk 15 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/New.hs]: done

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/New.hs (hunk 16 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/New.hs]: done

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Schema.hs (hunk 17 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/DB/Sqlite/Schema.hs]: done

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Schema.hs (hunk 18 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/DB/Sqlite/Schema.hs]: done

File: lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/Old.hs (hunk 20 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/Old.hs]: done

File: lib/wallet/src/Cardano/Wallet/DB/Store/WalletState/Migration.hs
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/DB/Store/WalletState/Migration.hs:1]: done

**Review by HeinrichApfelmus — APPROVED** [review]

File: lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs [hunk 21]
```diff
@@ -137,12 +140,13 @@ mkSharedStateFromRootXPrv
...
     -> ClearCredentials k
     -> Index 'Hardened 'AccountK
+    -> ChangeAddressMode
     -> AddressPoolGap
     -> ScriptTemplate
     -> Maybe ScriptTemplate
     -> SharedState n k
-mkSharedStateFromRootXPrv kF (RootCredentials rootXPrv pwd) accIx =
-    mkSharedStateFromAccountXPub kF accXPub accIx
+mkSharedStateFromRootXPrv kF (RootCredentials rootXPrv pwd) accIx mode =
```
Comment (paweljakubas) [inline lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs]: probably `mode` is unnecessary here and below

File: specifications/api/one-change-address-mode.md [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+## Description
+
+When balancing a transaction, the wallet selects
+unspent transaction outputs (UTxO) to
+cover the intended payments and the transaction fee.
+Any excess funds will be sent back to a *change address*
+which belongs to the wallet.
+
+By default, the Shelley-, Shared-, and Byron-style wallets
+create new a change address for every new transaction
```
Comment (paweljakubas) [inline specifications/api/one-change-address-mode.md]:
```suggestion
create a new change address for every new transaction
```

**Pull request merged**